src/trs/lexicon.rs

```diff
@@ -1,162 +1,182 @@
+    pub fn sample_term(
+        &mut self,
+        schema: &TypeSchema,
+        ctx: &mut Context,
+        invent: bool,
+        max_d: usize,
+    ) -> Result<Term, SampleError> {
+        let mut lex = self.0.write().expect("poisoned lexicon");
+        lex.sample_term(schema, ctx, invent, max_d, 0)
     }
-}
-impl ::std::error::Error for TypeError {
-    fn description(&self) -> &'static str {
-        "type error"
+    /// Sample a `Rule`.
+    pub fn sample_rule(
+        &mut self,
+        schema: &TypeSchema,
+        ctx: &mut Context,
+        invent: bool,
+        max_d: usize,
+    ) -> Result<Rule, SampleError> {
+        let mut lex = self.0.write().expect("poisoned lexicon");
+        lex.sample_rule(schema, ctx, invent, max_d, 0)
     }
-}
-
-#[derive(Debug, Clone)]
-pub enum SampleError {
-    TypeError(TypeError),
-    DepthExceeded(usize, usize),
-    OptionsExhausted,
-    Subterm,
-}
-impl From<TypeError> for SampleError {
-    fn from(e: TypeError) -> SampleError {
-        SampleError::TypeError(e)
+    /// Give the log probability of sampling a Term.
+    pub fn logprior_term(
+        &self,
+        term: &Term,
+        schema: &TypeSchema,
+        ctx: &mut Context,
+        invent: bool,
+    ) -> Result<f64, SampleError> {
+        let lex = self.0.read().expect("poisoned lexicon");
+        lex.logprior_term(term, schema, ctx, invent)
     }
-}
-impl From<polytype::UnificationError> for SampleError {
-    fn from(e: polytype::UnificationError) -> SampleError {
-        SampleError::TypeError(TypeError::Unification(e))
+    /// Give the log probability of sampling a Rule.
+    pub fn logprior_rule(
+        &self,
+        rule: &Rule,
+        schema: &TypeSchema,
+        ctx: &mut Context,
+        invent: bool,
+    ) -> Result<f64, SampleError> {
+        let lex = self.0.read().expect("poisoned lexicon");
+        lex.logprior_rule(rule, schema, ctx, invent)
+    }
+    /// Give the log probability of sampling a TRS.
+    pub fn logprior_trs(
+        &self,
+        trs: &TRS,
+        schemas: &[TypeSchema],
+        p_rule: f64,
+        ctx: &mut Context,
+        invent: bool,
+    ) -> Result<f64, SampleError> {
+        let lex = self.0.read().expect("poisoned lexicon");
+        lex.logprior_trs(trs, schemas, p_rule, ctx, invent)
+    }
+
+    /// merge two `TRS` into a single `TRS`.
+    pub fn combine(
+        &self,
+        rs1: &RewriteSystem,
+        rs2: &RewriteSystem,
+    ) -> Result<RewriteSystem, TypeError> {
+        assert_eq!(rs1.lex, rs2.lex);
+        let mut new_rs = rs1.clone();
+        // because the lexicon is the same, we only need to transfer rules
+        let mut new_rules = rs2.trs.rules.clone();
+        new_rs.trs.rules.append(&mut new_rules);
+        // update the context
+        let mut ctx = Context::default();
+        {
+            let lex = rs1.lex.0.read().expect("poisoned lexicon");
+            lex.infer_trs(&new_rs.trs, &mut ctx)?;
+        }
```

Comment:
Very similar to, say, `RewriteSystem::new`. Is there something to refactor there?

Reply:
Good catch!